File: toys/stuff/cpp/tools/markov.cpp
#include <iostream>
#include <deque>
#include <vector>
#include <string>
#include <map>

using std::cin;
using std::cout;
using std::istream;
using std::deque;
using std::vector;
using std::string;;
using std::map;

typedef deque<string> Prefix;
map<Prefix, vector<string>> statetab;
char NONWORD[] = "\n";

#define MAXGEN 10000
#define NPREF  2

void add(Prefix& prefix, const string& s)
{
    if (prefix.size() == NPREF) {
        statetab[prefix].push_back(s);
        prefix.pop_front();
    }
    prefix.push_back(s);
}

void build(Prefix& prefix, istream& in)
{
    string buf;
    while (in >> buf)
        add(prefix, buf);
}

void generate(int nwords) 
{
    Prefix prefix;
    int i;

    for (i = 0; i < NPREF; i++)
        add(prefix, NONWORD);

    for (i = 0; i < nwords; i++) {
        vector<string>& suf = statetab[prefix];
        const string& w = suf[rand() % suf.size()];
        if (w == NONWORD)
            break;
        cout << w << "\n";
        prefix.pop_front();
        prefix.push_back(w);
    }
}

int main(void)
{
    int nwords = MAXGEN;
    Prefix prefix;

    for (int i = 0; i < NPREF; i++)
        add(prefix, NONWORD);
    build(prefix, cin);
    add(prefix, NONWORD);
    generate(nwords);
    return 0;
}

File: toys/misc/java/Markov.java
import java.util.*;
import java.io.*;

class Prefix {
    public Vector pref;
    Prefix(int n, String str) {
        pref = new Vector();
        for (int i = 0; i < n; ++i)
            pref.addElement(str);
    }
    Prefix(Prefix p) {
        pref = (Vector) p.pref.clone();
    }
    public boolean equals(Object o) {
        Prefix p = (Prefix) o;
        for (int i = 0; i < pref.zize(); ++i) {
            if (!pref.elementAt(i).equals(p.pref.elementAt(i)))
                return false;
        }
        return true;
    }
}

class Chain {
    static final int NPREF = 2;
    static final String NONWORD = "\n";
    Hashtable statetab = new Hashtable();
    Prefix prefix = new Prefix(NPREF, NONWORD);
    Random rand = new Random();
    void add(String word) {
        Vector suf = (Vector) statetab.get(prefix);
        if (suf == null) {
            suf = new Vector();
            statetab.put(new Prefix(prefix), suf);
        }
        suf.addElement(word);
        prefix.pref.removeElementAt(0);
        prefix.pref.addElement(word);
    }

    void build(InputStream in) {
        StreamTokenizer st = new StreamTokenizer(in);
        try {
            st.resetSyntax();
            st.wordChars(0, Character.MAX_VALUE);
            st.whitespaceChars(0, ' ');
            while (st.nextToken() != st.TT_EOF)
                add(st.sval);
            add(NONWORD);
        } catch (IOException e) {
            System.out.println(e);
        }
    }
    void generate(int nwords) {}
}


public class Markov {
    static final int MAXGEN = 10000;
    public static void main(String[] args) {
        Chain chain = new Chain();
        int nwords = MAXGEN;
        chain.build(System.in);
        chain.generate(nwords);
        System.out.println("hello, world");
    }
}
